runtime/src/staking.rs

```diff
@@ -0,0 +1,161 @@
+        // session keys
+        let sesh_key = <T as pallet_session::Config>::Keys::decode(&mut &rotate_keys[..])
+            .expect("SessionKeys decoded successfully");
+        pallet_session::Pallet::<T>::set_keys(
+            RawOrigin::Signed(controller_account.clone()).into(),
+            sesh_key,
+            vec![],
+        )?;
+
+        let validator_prefs = pallet_staking::ValidatorPrefs {
+            commission: Perbill::from_percent(0),
+            ..Default::default()
+        };
+
+        // applying for the role of "validator".
+        StakingPallet::<T>::validate(
+            RawOrigin::Signed(controller_account.clone()).into(),
+            validator_prefs,
+        )?;
+
+        Ok(().into())
+    }
+
+    fn lock_extra_for_staking(
+        stash_account: &StakeAccountId<T>,
+        controller_account: &StakeAccountId<T>,
+        value: u128,
+    ) -> DispatchResultWithPostInfo {
+        // get the stash account first
+        if let Some(stake_ledger) = <StakingPallet<T>>::ledger(controller_account.clone()) {
+            if stash_account != &stake_ledger.stash {
+                log::error!(
+                    "stash account {:?} has no permission to stake.",
+                    stash_account
+                );
+                return Err(pallet_utxo::Error::<T>::NoPermission)?;
+            }
+
+            StakingPallet::<T>::bond_extra(
+                RawOrigin::Signed(stake_ledger.stash).into(),
+                value.into(),
+            )?;
+            return Ok(().into());
+        }
+
+        log::error!("check sync with pallet-staking.");
+        return Err(pallet_utxo::Error::<T>::InvalidOperation)?;
+    }
+
+    fn unlock_request_for_withdrawal(
+        stash_account: &StakeAccountId<T>,
+    ) -> DispatchResultWithPostInfo {
+        // get the controller account, given the stash_account.
+        let controller_account = <StakingPallet<T>>::bonded(stash_account.clone())
+            .ok_or(pallet_utxo::Error::<T>::StashAccountNotFound)?;
+
+        // stop validating / block producing
+        StakingPallet::<T>::chill(RawOrigin::Signed(controller_account.clone()).into())?;
+
+        // get the total balance to free up
+        let stake_ledger = <StakingPallet<T>>::ledger(controller_account.clone())
+            .ok_or(pallet_utxo::Error::<T>::ControllerAccountNotFound)?;
+
+        // unbond
+        StakingPallet::<T>::unbond(
+            RawOrigin::Signed(controller_account).into(),
+            stake_ledger.total,
+        )?;
+
+        Ok(().into())
+    }
+
+    fn withdraw(stash_account: &StakeAccountId<T>) -> DispatchResultWithPostInfo {
+        // get the controller account, given the stash_account.
+        let controller_account = <StakingPallet<T>>::bonded(stash_account.clone())
+            .ok_or(pallet_utxo::Error::<T>::StashAccountNotFound)?;
+
+        let stake_ledger = <StakingPallet<T>>::ledger(controller_account.clone())
```

Comment:
Please explain what's going on here. You're using the size of `unlocking` to figure out the current state. Please explain how you do that above the thing. Even better, move it to its own function and return an enum value representing your conclusion.